### PR TITLE
Add JSP (Java Server Pages) lexer

### PR DESCRIPTION
**Pull request by miparnisari**
Tested with close to 300 JSP files 😄 
image

image

### CODE REVIEW

File: lib/rouge/lexers/jsp.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+            state :jsp_directive2 do
+                rule /(\/>)/, Name::Tag, :pop!
+
+                mixin :tag
+            end
+
+            state :jsp_expression do
+                rule(/<\/jsp:(#{actions.join('|')})>/) { token Name::Tag; pop! }
+                rule(/[^<\/]+/) { delegate @java }
```
Comment (dblessing) [inline lib/rouge/lexers/jsp.rb]: Although there's technically nothing wrong with this syntax, let's stick to `do ... end` style blocks. I don't see this style used anywhere else in the code so let's stay consistent.

Comment (jneen) [reply lib/rouge/lexers/jsp.rb]: ??? i use this style all the time...

File: lib/rouge/lexers/jsp.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                rule(/<jsp:directive\.(#{directives.join('|')})/) { token Name::Tag; push :jsp_directive2 }
+
+                rule(/<jsp:(#{actions.join('|')})>/) { token Name::Tag; push :jsp_expression}
+
+                rule(/<[a-zA-Z]*:[a-zA-Z]*/) { token Name::Tag; push :jsp_tag_start }
+
+                rule(/<\/[a-zA-Z]*:[a-zA-Z]*/) { token Name::Tag; push :jsp_tag_end }
+
+                rule(/<%[!=]?/) { token Name::Tag; push :jsp_expression2 }
+            end
```
[the remark this answers is not in the document]
Comment (jneen) [reply lib/rouge/lexers/jsp.rb]: the parentheses are required in these cases?

[the remark this answers is not in the document]
Comment (jneen) [reply lib/rouge/lexers/jsp.rb]: but yes, the non-block style would be better.

**Comment by jneen** [comment]
For template languages like this, it is more appropriate to delegate to the parent language than to subclass it (see other template lexers like ERB for example), since the templating system acts as a preprocessor *before* the html syntax is interpreted.

File: lib/rouge/lexers/jsp.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (dblessing) [reply lib/rouge/lexers/jsp.rb]: Derp, thanks.

File: lib/rouge/lexers/jsp.rb (hunk 1 above)
Comment (dblessing) [reply lib/rouge/lexers/jsp.rb]: Ah, yep. I see it now. Don't know why I wasn't finding these examples earlier.

**Comment by dblessing** [comment]
Thanks for pointing out the ERB example @jneen. That looks nice and clean. @miparnisari Can you take a look, please, and see how that might work?

**Comment by miparnisari** [comment]
Thanks @jneen and @dblessing ! I'll have a look at the ERB example later.

**Comment by miparnisari** [comment]
@jneen @dblessing i've updated the lexer to subclass it from `TemplateLexer`.

**Comment by dblessing** [comment]
Thanks @miparnisari I'll review shortly.

**Comment by miparnisari** [comment]
Both done.

**Comment by miparnisari** [comment]
Any chance of releasing this soon in version 3.2.2? @dblessing

**Pull request merged**